### ISSUE TITLE
Factor "identity" transfer encoding support out of HTTPChannel.rawDataReceived into an object like the chunked transfer encoding object

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3605]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3605|
|Type|enhancement|
|Created|2009-01-08 00:16:58Z|
|Branch|https://github.com/twisted/twisted/tree/identity-transfer-encoding-3605|

This will simplify `rawDataReceived` and make the code re-usable by the HTTP client.

<details><summary>Searchable metadata</summary>

```
trac-id__3605 3605
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_identity_transfer_encoding_3605 branches/identity-transfer-encoding-3605
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1231373818000000 1231373818000000
changetime__1231772442000000 1231772442000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [25869]) Branching to 'identity-transfer-encoding-3605'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

Please review.

The `noMoreData` methods may seem like a strange addition since there are no callers.  However, the HTTP client will call this method.  Since I developed this code in the client branch, I'd already written them and the tests.  It seems like a waste of time to try to split up this code any more.  Once the HTTP client code is merged, there will be callers.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to @exarkun|
|-|-|

I just have one question: why making `_DataLoss` private and `PotentialDataLoss` public? Can you add a small comment in the source to explain that?

Otherwise, this is nice, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [25946]) expand _DataLoss docstring

refs [#3605](https://github.com/twisted/twisted/issues/3605)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [25947]) expand PotentialDataLoss docstring a bit

refs [#3605](https://github.com/twisted/twisted/issues/3605)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [25948]) Merge identity-transfer-encoding-3605

Author: exarkun, itamar
Reviewer: therve
Fixes: [#3605](https://github.com/twisted/twisted/issues/3605)
Refs: [#10610](https://github.com/twisted/twisted/issues/10610)

Add an identity transfer-encoding decoder, similar to the chunked transfer-encoding
decoder, to `twisted.web.http` and use it to simplify `HTTPChannel` slightly by
removing the knowledge of any particular transfer-encoding from its request body
handling code.

Also expand the transfer-encoding decoder interface so that it will be useful for
an HTTP client and rename the chunked decoder so that its name more closely reflects
what it does.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
